### DISCUSSION TITLE
Cascadia Code 2404.23

### DISCUSSION THREAD

**Discussion by DHowett** [microsoft/cascadia-code#728] [Announcements]
I don't think I could say anything to top this one, so here's the release notes for the first version of Cascadia in three years!
 
### Symbols for Legacy Computing (and more!)

Thanks to the tireless work of @PhMajerus, the Cascadia family now supports a whopping 1140 new glyphs covering sextants, octants, large type pieces, eights, sedecimants, quadrants (separated), segmented digits, circles and checkerboards. In addition, the existing block elements have been aligned to fit the same grid as the new characters to make for seamless ANSI art.

![Demo screen (2xDPI)](https://github.com/microsoft/cascadia-code/assets/189190/52d6a10a-7e33-40fd-9df3-00fe85003c3c)

See #708, #721, #723 and #727 for more details.

### "[Nerd Fonts](https://nerdfonts.com/)"

In addition to all of the above characters, the Cascadia family now comes with a native "Nerd Font" variant! It includes the entire set of glyphs as of April 2024 (all 9209 of them), and supersedes both unofficial variants "Caskaydia Cove" and "Delugia Code". Every glyph is metrics-compatible with the rest of Cascadia, so there should be no (or very few?) visual aberrations.

![image](https://github.com/microsoft/cascadia-code/assets/189190/ec449963-f524-4e01-bbea-936ff5f628ff)

See #720 for all the details!

<hr /><em>This discussion was created from the release <a href='https://github.com/microsoft/cascadia-code/releases/tag/v2404.23'>Cascadia Code 2404.23</a>.</em>

**Comment by totkeks** [comment]
Very cool, thank you. I was watching Scott Hanselman on YouTube the other day and he talked about the Cascadia Code version with built-in NerdFonts. Then I came here, and didn't see a new release for years. But today you delivered it. Thanks 👍😊

**Comment by ShayBox** [comment]
Is it possible to use the Cursive `ss01` variant without using the OpenType variant now for editors like JetBrains? The freeze project listed in the README doesn't work anymore.

**Reply by DHowett** [reply]
As part of this release, we have made no changes to JetBrains' software.

Are you getting an error message from `fonttools-opentype-feature-freezer`? We haven't meaningfully changed the OpenType variant of the font.

**Reply by ShayBox** [reply]
I tried to use OpenType feature freezer, but it was only able to freeze regular italics, it didn't freeze the `ss01` variant when I tried to use it. VS Code is the only editor or program I've used that supports OpenType variants.

EDIT: Every time I've tried to use the Cascadia Code cursive variant for things like JetBrains IDEs, Terminals, or photo editors like Gimp or Krita, I've been unable to.

**Comment by junguler** [comment]
nice font, great release too, good job

**Comment by sonulohani** [comment]
Loving the new release, great font for programming and terminal

**Comment by agpt8** [comment]
So, which one is recommended to install? I believe the powerline (PL variant) has an overlap with the nerd font (NF) variant, so is the PL variant needed at all now? If these two are separate, shouldn't they be combined to have just the normal and the NF variant?

**Reply by aaronbell** [reply]
My understanding is that not everyone wants a full NF font, and just want PL. No harm in providing more options if possible, IMO.

**Comment by stanio** [comment]
>   supersedes both unofficial variants "Caskaydia Cove" and "Delugia Code"

I still use the "Caskaydia Cove" for its [non-monospaced](https://github.com/ryanoasis/nerd-fonts?tab=readme-ov-file#features) (not to be confused with the proportional) variant where glyphs may span two character cells.  Many glyphs of the strict monospaced variant, such is this official "Cascadia Code" release, end up too tiny for my preference.

**Reply by aaronbell** [reply]
Fair enough. I plan to provide alternate versions of the NF in the future.

**Reply by varun701** [reply]
NerdFonts have 3 versions if the source font is Monospace
NFM  (Nerd Font Mono) - Icons are monospaced (Does not look good for many symbols)
NF  (Nerd Font)  - Icons are not monospaced (Monospace grid preserved - blanks after icons are needed) 
NFP (Nerd Font Propo) - Icons are not monospaced (Monospace grid destroyed - no blanks after symbols needed)

All 3 may have NL (No Ligature) Alternate versions of them - making it total of 6 versions

NerdFonts you provided has only two of them
> NFM (Cascadia Code NF)
> NL NFM (Cascadia Mono NF)

I usually use all six versions for different use cases
NFP & NL NFP as Interface font
NF & NL NF as Text font
NFM & NL NFM as Code/Monospaed font

I uninstalled caskaydia cove just to realise this release does not have everything I need.

It would be great if you do provide alternate verisons. However what I wanted to point was naming of fonts.
This release has a `Caskaydia Cove NFM` version named `Cascadia Code NF` which you may want to change in future releases or put a disclaimer of using a different name than other NerdFonts.

**Comment by PeterAdam** [comment]
"Note: If you have previously installed a version of Cascadia Code, please uninstall the previous version prior to installing a new version. Not doing so can result in improper rendering."

Tried to remove, but the new Settings app says:

"Are you sure? This font is part of the following application: Windows Terminal To remove the font, you have remove the application."

I hope I do not have to reinstall Windows for the new release. /s

**Reply by aaronbell** [reply]
You will likely get the new version once Windows Terminal updates. Else, you can install and use the NF / PL versions.

**Comment by charlesroper** [comment]
What is the recommended "default" set to install? OTF static, TTF static, or TTF variable?

**Reply by charlesroper** [reply]
Found the info I needed in the readme: https://github.com/microsoft/cascadia-code?tab=readme-ov-file#font-formats

**Reply by nashpatty** [reply]
I had a similar question, and a follow up question. @DHowett @aaronbell is OTF the recommended format to install on a mac? In the latest version of the font, I'm still seeing weird visual discrepancies in the TTF versions (I mentioned them in this issue: [#702](https://github.com/microsoft/cascadia-code/issues/702)). Most glyphs have have odd anchors that are visible.

**Reply by aaronbell** [reply]
I generally recommend variable TTF for any OS. I use the TTF on my Mac and don't observe the level of visual discrepancies that you do. Essentially the problem is in the OS / application text rasterizing engine (see #350) and I'm not aware of any way to resolve it on my end for the variable font. I think when I refactor the build system, I'll make sure to remove overlaps on the static TTFs, which should solve it there, but will result in requiring me to dump the hinting, so it'll render blurrier. Tradeoffs :(

**Reply by nashpatty** [reply]
Thanks for linking that issue, seems like clearing the font cache on my mac resolved the visual bugs in the static TTFs, at least. The reason why I found it odd was because these discrepancies were also visible in font book while installing the font on my mac. Could I ask if you have system wide font smoothing enabled or disabled on your mac environment?

**Reply by aaronbell** [reply]
Oh interesting. I didn't realize that the cache would affect rendering. 
No, I don't have any special modifications to my system's font smoothing at present.

**Reply by hamburnyog** [reply]
I've been wondering about this as well. Nice insights guys, thanks!

**Comment by danielrippa** [comment]
On Windows 10 22H2 after a fresh manual install of this version (2404.023 as per the Font Properties in WIndows Settings), I don't see any of the new glyphs (I'm mainly interested in Large Type Pieces, Quadrants and friends, etc). I don't see them in CharMap, I can't get to print them using their Unicode codepoints, etc.

**Reply by PhMajerus** [reply]
AFAIK, charmap does not offer the new characters yet. I'm not sure what the process would be at Microsoft and if the Terminal/Cascadia team has any contact with the team in charge of charmap to get it updated. Also remember those characters are to be officially added in Unicode 16, which is only planned to be released next month.

You can input the characters using the Alt key followed by their code point. I believe you must use hexadecimal for those characters, their decimal values are not recognized.
So the first large type piece for example would be input by holding the `Alt` key, and typing `+1ce1a`.
Unfortunately, some apps use some of these Alt combinations as menu items shortcuts (accelerators), so they might get interrupted and handled differently.

Another solution is to simply copy/paste them from somewhere. For example, you can use my _cheat sheet_ for quandrants, sextants, and octants by typing `curl https://raw.githubusercontent.com/PhMajerus/Documents/main/CheatSheets/Blocks%20tables.txt` in a terminal and copy them from the chart displayed.
For the large type pieces, I don't have a _cheat sheet_, but you can copy them from my _how to_ document using `curl https://raw.githubusercontent.com/PhMajerus/Documents/main/HowTos/HowTo%20Large%20Type%20Pieces.txt`.

**Reply by danielrippa** [reply]
I can confirm that this works in Windows Terminal, but, sadly, not in cmd.
Thanks for the quick response!

**Reply by PhMajerus** [reply]
I added a cheat sheet for the large type pieces to make it easier to quickly copy/paste them:
`curl https://raw.githubusercontent.com/PhMajerus/Documents/main/CheatSheets/Large%20Type%20Pieces.txt`

![image](https://github.com/user-attachments/assets/a3dca33c-8958-41e8-ac3c-662cbb4538a8)

**Reply by danielrippa** [reply]
Thanks for this.
I would also appreciate your suggestions regarding the missing glyphs in the different groups.  It becomes hunting and pecking otherwise. Thanks in advance!

**Reply by PhMajerus** [reply]
The Symbols for Legacy Computing and Symbols for Legacy Computing Supplement cover 938 new characters, you can't really just consider them a single group.
I focused on pseudopixels, related diagonals and circular shapes, segmented digits, and large type pieces, adding 576 new characters, almost all of which required two glyph versions for a total of 1140 new glyphs:

`U+1FB95`, `U+1FB96` : Checker Board Fill and Inverse Checker Board Fill (full block size)
`U+2427` - `U+2429` : New Control Pictures ("delete" checkerboards)
`U+1FBE4` - `U+1FBE7` : Centered Quadrants
`U+1FBE8` - `U+1FBEF` : Circular Quadrant characters
`U+1FB68` - `U+1FB6F` : Triangular Quadrant characters
`U+1FB00` - `U+1FB3B` : Sextant characters (2×3 mosaics)
`U+1FB3C` - `U+1FB67` : Diagonal Sextant characters (aka smooth sextants)
`U+1CD00` - `U+1CDE5` : Octants (2×4 mosaics) 
`U+1CE90` - `U+1CEAF` : "Sedecimants" (4×4 mosaics)
`U+1FB70` - `U+1FB80`, `U+1FB82` - `U+1FB8B` : "Eights" (8×8 patterns)
`U+1FB97`, `U+1FBCE`, `U+1FBCF` : Misc. blocks
`U+1CC21` - `U+1CC2F` : Separated quadrants (2×2)
`U+1CE51` - `U+1CE8F` : Separated sextants (2×3)
`U+1FBF0` - `U+1FBF9` : Segmented digits (LED/LCD display style)
`U+1CE1A` - `U+1CE50` : Large Type Pieces

Here they are in Windows Terminal:
![image](https://github.com/user-attachments/assets/d3475d1f-6abf-4460-a4e2-5b4b53c08b72)

There are other characters introduced with Unicode 16.0 Symbols for Legacy Computing and Supplement that I'd like to add later, such as the Apple II MouseText, new box drawing characters (lines), shaded blocks, outlined letters and digits.
The others are mostly sprites for games on legacy systems that I don't think will ever be used in a terminal, so I don't think we'll be adding all of the 362 other characters anytime soon.
I think we should focus on bringing characters that came from legacy terminals, such as HP 2640 Series (#715), and then from legacy computers such as the Commodore 64, Apple II, and TRS-80.

**Reply by PhMajerus** [reply]
Just like for the large type pieces, I have _cheat sheets_ available to make it easy to copy/paste most of those characters:

`curl https://raw.githubusercontent.com/PhMajerus/Documents/main/CheatSheets/Blocks%20tables.txt`
![image](https://github.com/user-attachments/assets/db3860d2-047c-4f97-88ba-646a5c6098e1)

`curl https://raw.githubusercontent.com/PhMajerus/Documents/main/CheatSheets/More%20blocks%20tables.txt`
![image](https://github.com/user-attachments/assets/3b705009-95d8-4622-b949-fc72eb8d6398)

`curl https://raw.githubusercontent.com/PhMajerus/Documents/main/CheatSheets/Lines%20tables.txt`
![image](https://github.com/user-attachments/assets/96046a5d-9a3d-4319-8f86-9f0266993e81)

Note these show some missing `�` characters, as they already include some other characters I'd like to add later.

**Reply by danielrippa** [reply]
Sorry, I missed these messages.
Thank you! I really appreciate the time you take and the very detailed responses!

**Comment by jpawlowski** [comment]
For those interested:

I've been developing a [PowerShell installer for Nerd Fonts](https://www.powershellgallery.com/packages/Invoke-NerdFontInstaller). Given that _Cascadia Code_ and _Cascadia Mono_ now have official Nerd Font and Powerline variants, I've decided to add them to the selection. You can also opt to install the traditional variants if you prefer (particularly appealing to macOS and Linux users, since Windows 11 includes the Windows Terminal so that font files will be automatically excluded by the script).

To make it even easier for individuals to install the extra files, the script can be executed directly from the web using a simple one-liner:

```powershell
& ([scriptblock]::Create((iwr 'https://to.loredo.me/Install-NerdFont.ps1')))
```

![PowerShell Nerd Fonts Installer](https://raw.githubusercontent.com/jpawlowski/nerd-fonts-installer-PS/main/images/nerd-fonts-installer.gif)

A nice menu will let you select the font(s) that you want and install them, without any other local dependencies. You even get the links to the preview pages that you might be used to when visiting [www.nerdfonts.com](https://www.nerdfonts.com/).

Regarding security, I am aware that executing a script directly from the web can be risky. However, this practice is common in various contexts, and I believe that if the source is trustworthy and employs transport layer encryption, one might consider the convenience to outweigh the risk.

Nevertheless, please use it at your own risk; it is offered "as is." Naturally, you are welcome to review the source code and contribute pull requests on GitHub.

**Comment by mat48** [comment]
hey there hard working fellows :)
I'm new  here, Just wanted to note that after watchin some of the bugfixes and features in the release section, I have to admit that there are a lot of amazing work getting done in places that most people wont notice !

mentioning the changes in feature/bug fixes section brings more attention to the great details of this job. thank you and all the good community members who are helping people with their feedbacks or any other way :) <3

**Comment by danielrippa** [comment]
Not sure if this is the best place to ask but... here I go anyway.
Is there some (say, win32 api) function that I can use to determine how many characters cells a certain character will take?
I see some take just one character cell, others take two.
An image speaks a thousand words.
![image](https://github.com/user-attachments/assets/643c67f1-3f2f-416f-a3b1-6ace27d214af)
![image](https://github.com/user-attachments/assets/10203297-123f-4a2f-bf81-b13d230a8272)
Building a database of cell widths for each specific font is overkill.

**Reply by alexeyten** [reply]
Cascadia doesn't have these characters.

**Reply by danielrippa** [reply]
As a workaround, I'll temporarily print the character whose width I need to measure to a separate screen buffer. I'll then print another character next to it and observe its position to infer the width of the target character. *sigh*

**Reply by danielrippa** [reply]
Easier said than done, though.
https://github.com/FarGroup/FarManager/issues/873#issuecomment-2351484287

**Reply by PeterAdam** [reply]
https://learn.microsoft.com/en-us/windows/win32/gdi/character-widths
https://learn.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-gettextextentpoint32a?redirectedfrom=MSDN

**Reply by danielrippa** [reply]
Great! Thank you!

**Comment by brettmillerb** [comment]
Is there a reason the file size of the zip had jumped so dramatically over the past few releases?

It has gone from 23Mb to 143Mb

**Reply by mfocko** [reply]
Yes. Cascadia release contains cartesian product of

    { Mono, Code } x { n/a, PL, NF }

- Mono — classic monospaced font
- Code — with enabled ligatures
- n/a — classic font
- PL — additionally powerline characters
- NF — PL + Nerd Font characters

since the second to last release they ship in one ZIP all combinations; NF has been added, takes up most of the space, you can have a look at https://www.nerdfonts.com/ how many glyphs are added :)

And also they ship all possible formats, so apart from the difference in the font itself there’s also:

    { ttf, otf, woff2 } x { static, variable }

so in the end you get one ZIP-file of

    { ttf, otf, woff2 } x { static, variable } x { Mono, Code } x { n/a, PL, NF }

For desktop you usually need only one of the ttf/otf, also it would be recommended to install both static and variable versions (as some applications **still** (_sigh_) have issues with variable fonts), as for the ligatures/no-ligatures and pure font/powerline/nerdfont glyphs, that’s your personal preference.

---

For example, comparing the variable ttf font (Cascadia Code) itself and PL with NF variants you get:

- font itself — 721 KiB
- PL — 747 KiB (+3.54%)
- NF — 2.4 MiB (+233.58% to PL, +245.37% to pure font)
